include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -1683,16 +1691,21 @@ FEPointEvaluation<n_components_, dim, spacedim, Number>::prepare_evaluate_fast(
     }
   for (unsigned int comp = 0; comp < n_components; ++comp)
     {
-      const unsigned int *renumber_ptr =
-        renumber.data() +
-        (component_in_base_element + comp) * dofs_per_component;
       if (use_face_path)
         {
           ScalarNumber *input  = scratch_data_scalar.begin();
           ScalarNumber *output = input + dofs_per_component;
 
-          for (unsigned int i = 0; i < dofs_per_component; ++i)
-            input[i] = solution_values[renumber_ptr[i]];
+          if (renumber.empty())
+            input = const_cast<ScalarNumber *>(solution_values.data());
+          else
```

Comment:
Can you avoid the `const_cast` and declare `input` as `const ScalarNumber` instead? In the `else` branch below, we should then set `scratch_data_scalar[i] = solution_values[renumber_ptr[i]];` and afterwards assign `input` to it.